File: src/main/java/com/hileco/cortex/analysis/processors/KnownLoadProcessor.java
package com.hileco.cortex.analysis.processors;

import com.hileco.cortex.analysis.Graph;
import com.hileco.cortex.instructions.debug.NOOP;
import com.hileco.cortex.instructions.io.LOAD;
import com.hileco.cortex.instructions.stack.PUSH;
import com.hileco.cortex.vm.ProgramStoreZone;
import lombok.Value;

import java.math.BigInteger;
import java.util.Map;

@Value
public class KnownLoadProcessor implements Processor {
    private final Map<ProgramStoreZone, Map<BigInteger, BigInteger>> knownData;

    @Override
    public void process(Graph graph) {
        graph.getGraphBlocks().forEach(graphBlock -> graphBlock.getGraphNodes().stream()
                .filter(graphNode -> graphNode.isInstruction(LOAD.class))
                .filter(graphNode -> graphNode.hasOneParameter(0, parameter -> parameter.isInstruction(PUSH.class)))
                .forEach(loadGraphNode -> {
                    var pushGraphNode = loadGraphNode.getParameters().get(0);
                    var load = (LOAD) loadGraphNode.getInstruction().get();
                    var push = (PUSH) pushGraphNode.getInstruction().get();
                    var address = new BigInteger(push.getBytes());
                    if (this.knownData.containsKey(load.getProgramStoreZone())) {
                        var knownDataMap = this.knownData.get(load.getProgramStoreZone());
                        if (knownDataMap != null && knownDataMap.containsKey(address)) {
                            pushGraphNode.getInstruction().set(new NOOP());
                            loadGraphNode.getInstruction().set(new PUSH(knownDataMap.get(address).toByteArray()));
                        }
                    }
                }));
    }
}

File: src/main/java/com/hileco/cortex/analysis/processors/KnownLoadProcessor.kt
package com.hileco.cortex.analysis.processors

import com.hileco.cortex.analysis.Graph
import com.hileco.cortex.instructions.debug.NOOP
import com.hileco.cortex.instructions.io.LOAD
import com.hileco.cortex.instructions.stack.PUSH
import com.hileco.cortex.vm.ProgramStoreZone
import java.math.BigInteger

class KnownLoadProcessor(private val knownData: Map<ProgramStoreZone, Map<BigInteger, BigInteger>>) : Processor {
    override fun process(graph: Graph) {
        graph.graphBlocks.forEach { graphBlock ->
            graphBlock.graphNodes.asSequence()
                    .filter { it.instruction is LOAD }
                    .filter { graph.edgeMapping.hasOneParameter(it, 0) { parameter -> parameter.instruction is PUSH } }
                    .forEach {
                        val pushGraphNode = graph.edgeMapping.parameters(it).first()
                        if (pushGraphNode != null) {
                            val load = it.instruction as LOAD
                            val push = pushGraphNode.instruction as PUSH
                            val address = BigInteger(push.bytes)
                            knownData[load.programStoreZone]?.let { knownDataMap ->
                                knownDataMap[address]?.let { knownData: BigInteger ->
                                    pushGraphNode.instruction = NOOP()
                                    it.instruction = PUSH(knownData.toByteArray())
                                }
                            }
                        }
                    }
        }
    }
}
